gui/default/syncthing/core/syncthingController.js

```diff
@@ -651,6 +652,17 @@ angular.module('syncthing.core')
             $scope.remoteNeedDevice = undefined;
         }
 
+        function setDefaultTheme(){
+          // If theme is default and there is no support for prefers-color-scheme
+          if ($scope.config.gui.theme === "default" && window.matchMedia('(prefers-color-scheme: dark)').media === 'not all') {
+            document.documentElement.style.display = 'none';
+            document.head.insertAdjacentHTML(
+              'beforeend',
+              '<link rel="stylesheet" href="/theme-assets/light/assets/css/theme.css" onload="document.documentElement.style.display = \'\'">'
+            );
```

Comment:
Why are we setting something to display none? Also, on a browser that does not support this, we'd be repeatedly injecting the same element on every config update (even if nothing has changed).

Reply:
Also, I guess this ends up being a third css file we load if there is no support?
Or does it not load the other two?
If it does not, cannot we just match on 'not all' in css?

Reply:
Or use a not(selector) ?

Reply:
> Why are we setting something to display none?

This is suggested by [web.dev](https://web.dev/prefers-color-scheme/#loading-strategy): "to avoid a [flash of unstyled content](https://en.wikipedia.org/wiki/Flash_of_unstyled_content), I hide the content of the page until light.css has loaded."

> we'd be repeatedly injecting the same element on every config update

my bad, I only tested loading a new theme and in that case the page is reloaded.

Reply:
> If it does not, cannot we just match on 'not all' in css? Or use a not(selector) ?

I tried not using JS and changing the light theme condition to `not all and (prefers-color-scheme: dark)`. It works correctly on Firefox 70 but it doesn't on Firefox 52 (a browser without support for `prefers-color-scheme`).

Reply:
I would expect 'not all' to be a value matcher for the pref?

Reply:
ok tried it myself, it seems you can't do it without JS.
 It seems that not all browsers support `mediaMatch`, so we should check if that exists on the window.

Reply:
> > If it does not, cannot we just match on 'not all' in css? Or use a not(selector) ?
> 
> I tried not using JS and changing the light theme condition to `not all and (prefers-color-scheme: dark)`. It works correctly on Firefox 70 but it doesn't on Firefox 52 (a browser without support for `prefers-color-scheme`).

`not all` means `not screen`, `not print`, and `not speech`, so it is basically the same as not having the media query at all. It has nothing to do with support for the specific feature, and of course will be completely ignored by any browser.

I am no Web development expert, but this honestly looks very confusing. Why not just apply the light theme always, unconditionally, just with CSS and drop the JavaScript altogether? This way you can be 100% sure that everything will work correctly in old browsers. The code will be much simpler too, with no additional JavaScript hackery.

Reply:
Because someone will add some rule (all td's are white) to the light theme, which will break dark theme, because both files are loaded, and dark theme doesn't explicitly state that all td's are dark, and inherit the colour from the body.

Reply:
But is this not the case right now (without this commit, that is)?

The original default theme was light, and the other themes have been built on top of it. Is this wrong? Do you want to completely separate the light theme from the rest? The problem is that by doing so, there will be no fallback, which means good bye to progressive enhancement.

On a side note, if that is the intention, this change will likely break any users' custom themes which until now could safely rely on the fact that that the default theme was light.

Reply:
That is not the case now, as all the default stuff comes from bootstrap itself.
The dark theme just overrides what bootstrap does, which we can assume is stable and we can cover the ground.

Our light theme on the other hand, is customisation to bootstrap, which for now does not have a dodgy td: white, but it might, at which point that will override the dark td inherited from body when setting body colour in bootstrap.

Reply:
Oh, you are right. I wrongly assumed that the `theme.css` from the `default` theme was applied to all the other themes, while this does not seem to be the case. I am sorry for the confusion.

I still do not believe that using JavaScript to detect browser features in order to provide a "fallback" for CSS is the right way. If something goes wrong with the JavaScript hackery here (which is very likely e.g. in a quirky older browser), then neither the light nor the dark theme will be applied at all. This may not be a big deal at the moment, as the two themes have a very limited set of styles right now, but may become a problem if they are expanded in the future.

Reply:
> `not all` means `not screen`, `not print`, and `not speech`, so it is basically the same as not having the media query at all. It has nothing to do with support for the specific feature, and of course will be completely ignored by any browser.

That media query can be confusing but it doesn't work that way. `not all and (prefers-color-scheme: dark)` is parsed as `not (all and (prefers-color-scheme: dark))`, that is, the negation of `prefers-color-scheme: dark`. 

`all` is included because one can't negate `prefers-color-scheme` directly; as [MDN says](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#Inverting_a_query's_meaning) : "The not keyword can't be used to negate an individual feature query, only an entire media query". See also [this SO answer](https://stackoverflow.com/a/36494725/3414720).

This query works on recent browsers (and it is the one used in the last revision of [Thunderbird's CSS](https://hg.mozilla.org/comm-central/file/tip/mail/themes/windows/mail/multimessageview.css#l23) to set a dark theme, see line 23), but it didn't work on Firefox 52 when I tried. You are welcome to suggest other possible media queries to see if we can come up with something compatible with older browsers, but sadly I don't think there is a pure CSS solution.

Reply:
I have updated myself on the media queries specification (https://www.w3.org/TR/css3-mediaqueries and https://www.w3.org/TR/mediaqueries-4), and yes, you are right about the `not` keyword, which is applied to the whole declaration following it. The whole `all` is not required though. By the official specification, you could even do a simple negation like this:

```
@import "/theme-assets/dark/assets/css/theme.css" screen and (prefers-color-scheme: dark);
@import "/theme-assets/light/assets/css/theme.css" not screen and (prefers-color-scheme: dark);
```

The problem is that this does not work in real life. While the syntax may work correctly in newer browsers, IE11 completely ignores it, despite the fact that theoretically the browser should support CSS3 media queries.

With this in mind, in my opinion, the only way to do the whole thing with CSS only is to apply one of the themes by default, and then override the values with the other one (i.e. use the CSS cascade). This would indeed require to maintain the two themes simultaneously. They could be even merged into a single file. If the Syncthing developers do not like this, then indeed a JavaScript band-aid may be the only possible method.